File: vfr-green-map/src/resources/tile-provider.ts
import { CachedProvider, CachedProviderRequest } from './cached-provider'
import { Point2d, Box2d } from "coordinates"
import { ITileReceiver } from "./i-tile-receiver";

export class TileProvider extends CachedProvider {
    private mapRoot: string;

    constructor(mapRoot: string) {
        super();
        this.mapRoot = mapRoot;
    }

    public retrieveTile(mapName: string, mapVersion: string, zoomLevel: number, location: Point2d, tileDimensions: Point2d ,
            receiver: ITileReceiver, data: any): void {
        let key = this.createKey(mapName, zoomLevel, location);
        let tileRequest = this.getExistingRequest(key);
        
        if ( tileRequest !== undefined && !tileRequest.isInError() ) {
            let tileRequestScoped = tileRequest as TileRequest;
            if (tileRequest.isLoaded()) {
                let image = tileRequestScoped.getImage();
                if ( image != null ) {
                    let region = new Box2d(0, 0, tileDimensions.x, tileDimensions.y);
                    receiver.receiveTile(location, region, image, data, true);
                }
            } else {
                tileRequestScoped.setReceiver(receiver, data);
                this.findTemporaryTile(mapName, zoomLevel, location, receiver, data);
            }
        } else {
            let url = `${this.mapRoot}/${mapName}_SEC_${mapVersion}/${zoomLevel}/${location.x}_${location.y}.png`;
            let newRequest = new TileRequest(this, receiver, location, tileDimensions, data, url);
            this.addRequestToQueue(key, newRequest);
            this.findTemporaryTile(mapName, zoomLevel, location, receiver, data);
        }
    }

    private createKey(mapName: string, zoomLevel: number, location: Point2d): string {
        return `${mapName}|${zoomLevel}|${location.x}_${location.y}`;
    }

    private findTemporaryTile(mapName: string, zoomLevel: number, requestLocation: Point2d, receiver: ITileReceiver, data: any): void {
        let divisor = 1;
        for ( let i = zoomLevel - 1; i >= 0; i-- ) {
            // Get the new cell that we want in the new zoom level
            divisor *= 2;
            let zoomLocation: Point2d = new Point2d(Math.floor(requestLocation.x / divisor), Math.floor(requestLocation.y / divisor));

            // See if that cell is available and draw it if it is
            let key = this.createKey(mapName, i, zoomLocation);
            let cachedRequest = this.getExistingRequest(key);
            if ( cachedRequest != undefined && cachedRequest.isLoaded() ) {
                // Get the bounds of that cell in the original zoom cells
                let originalZoomZoomLocation = new Point2d(zoomLocation.x * divisor, zoomLocation.y * divisor);
                let subsectionX = (requestLocation.x - originalZoomZoomLocation.x) / divisor;
                let subsectionY = (requestLocation.y - originalZoomZoomLocation.y) / divisor;
                let subsection = new Box2d(subsectionX, subsectionY, subsectionX + 1 / divisor, subsectionY + 1 / divisor);

                // Get the image and draw it
                let image = (cachedRequest as TileRequest).getImage();
                if ( image != null ) {
                    receiver.receiveTile(requestLocation, subsection, image, data, true);
                    return;
                }
            }
            //requestLocation = zoomLocation;
        }

        return;
    }
}

class TileRequest extends CachedProviderRequest {
    private tileImage: HTMLImageElement | null;
    private receiver: ITileReceiver | null;
    private location: Point2d | null;
    private dimensions: Point2d | null;
    private data: any | null;
    private url: string;

    constructor(provider: TileProvider, receiver: ITileReceiver, location: Point2d, dimensions: Point2d, 
        data: any, url: string) {
        super(provider);
        this.tileImage = null;
        this.receiver = receiver;
        this.location = location;
        this.dimensions = dimensions;
        this.data = data;
        this.url = url;
    }

    public sendRequest(): void {
        this.tileImage = new Image();
        this.tileImage.addEventListener("load", (event) => {
            let success = this.tileImage !== null && this.tileImage.complete && this.tileImage.naturalWidth > 1;
            this.completeRequest(success);
        });
        this.tileImage.src = this.url;
    }

    public broadcastData(): void {
        if (this.isLoaded() && !this.isInError() && 
            this.receiver != null && this.location != null && this.tileImage != null && this.dimensions != null) {

            let region = new Box2d(0, 0, this.dimensions.x, this.dimensions.y);
            this.receiver.receiveTile(this.location, region, this.tileImage, this.data, false);
            this.receiver = null;
            this.location = null;
            this.data = null;
        }
    }

    public setReceiver(receiver: ITileReceiver, data: any) {
        this.receiver = receiver;
        this.data = data;
    }

    public getImage(): HTMLImageElement | null {
        return this.tileImage;
    }
}

File: vfr-green-app/app/src/main/java/com/blueskycharts/app/map/resources/TileProvider.kt
package com.blueskycharts.app.map.resources

import android.content.Context
import android.graphics.Canvas
import com.blueskycharts.app.Constants
import com.blueskycharts.app.coordinates.Box2d
import com.blueskycharts.app.coordinates.Point2d
import com.blueskycharts.app.map.assetmanagement.TileAssetProvider
import com.blueskycharts.app.map.configuration.Inventory
import com.blueskycharts.app.map.view.Map
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import kotlin.math.floor

open class TileProvider(val context: Context, map: Map, private val group: Inventory.Group) : CachedProvider(map, 0), TileProviderInterface {
    val assetProvider
        get() = TileAssetProvider.getInstance(group)

    override fun retrieveTile(mapName: String, mapVersion: String, zoomLevel: Int, location: Point2d, tileDimensions: Point2d,
                     receiver: TileReceiver, data: Any?, canvas: Canvas ) {
        val key = this.createKey(mapName, zoomLevel, location);
        val tileRequest = this.getCachedItem(key);
        if ( tileRequest != null && tileRequest.loaded ) {
            val tileRequestScoped = tileRequest as TileRequest;
            if (tileRequestScoped.loaded) {
                tileRequestScoped.setReceiver(receiver, data);
                tileRequestScoped.broadcastData(true, canvas);
            } else {
                this.findTemporaryTile(mapName, zoomLevel, location, receiver, data, mapVersion, tileDimensions, canvas)
            }
        } else {
            this.findTemporaryTile(mapName, zoomLevel, location, receiver, data, mapVersion, tileDimensions, canvas)
            incrementAwaitingQueueAddition()
            GlobalScope.launch {
                try {
                    val newTileRequest = this@TileProvider.getExistingRequest(key);
                    if (newTileRequest != null && !newTileRequest.inError) {
                        val tileRequestScoped = newTileRequest as TileRequest;
                        tileRequestScoped.setReceiver(receiver, data);
                        if (tileRequestScoped.loaded) {
                            tileRequestScoped.broadcastData(false, canvas);
                        }
                    } else {
                        val newRequest = TileRequest(
                            this@TileProvider,
                            receiver,
                            data,
                            location,
                            tileDimensions,
                            zoomLevel,
                            mapName,
                            mapVersion
                        );
                        this@TileProvider.addRequestToQueue(key, newRequest);
                    }
                } finally {
                    decrementAwaitingQueueAddition()
                }
            }
        }
    }

    private fun createKey(mapName: String, zoomLevel: Int, location: Point2d): String {
        return "${mapName}|${zoomLevel}|${location.x.toInt()}_${location.y.toInt()}";
    }

    private fun findTemporaryTile(mapName: String, zoomLevel: Int, requestLocation: Point2d, receiver: TileReceiver, data: Any?,
        mapVersion: String, tileDimensions: Point2d, canvas: Canvas) {
        var divisor = 1;
        for ( i in zoomLevel - 1 downTo 0 ) {
            // Get the new cell that we want in the new zoom level
            divisor *= 2;
            val zoomLocation: Point2d = Point2d(floor(requestLocation.x / divisor), floor(requestLocation.y / divisor));

            // See if that cell is available and draw it if it is
            val key = this.createKey(mapName, i, zoomLocation);
            val cachedRequest = this.getCachedItem(key);
            if ( cachedRequest != null && cachedRequest.loaded ) {
                // Get the bounds of that cell in the original zoom cells
                val originalZoomZoomLocation = Point2d(zoomLocation.x * divisor, zoomLocation.y * divisor);
                val subsectionX = (requestLocation.x - originalZoomZoomLocation.x) / divisor;
                val subsectionY = (requestLocation.y - originalZoomZoomLocation.y) / divisor;
                val subsection = Box2d(subsectionX, subsectionY, subsectionX + 1 / divisor.toDouble(), subsectionY + 1 / divisor.toDouble());

                // Get the image and draw it
                val image = (cachedRequest as TileRequest).image;
                if ( image != null ) {
                    receiver.receiveTile(subsection, image, data, true, canvas);
                    return;
                }
            } else if ( cachedRequest == null ) {
                GlobalScope.launch {
                    val newRequest = TileRequest(this@TileProvider, receiver, data, zoomLocation, tileDimensions, i, mapName, mapVersion)
                    this@TileProvider.addRequestToQueue(key, newRequest);
                }
            }
        }
    }
}